File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Events/ExceptionInUpdatingCallbackCorruptionTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using System.Collections;
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4oUnit.Extensions.Fixtures;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Consistency;
using Db4objects.Db4o.Events;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Tests.Common.Events;

namespace Db4objects.Db4o.Tests.Common.Events
{
	/// <exclude></exclude>
	public class ExceptionInUpdatingCallbackCorruptionTestCase : AbstractDb4oTestCase
		, IOptOutMultiSession
	{
		public static void Main(string[] args)
		{
			new ExceptionInUpdatingCallbackCorruptionTestCase().RunSolo();
		}

		public class Holder
		{
			public IList list = new ArrayList();

			public ExceptionInUpdatingCallbackCorruptionTestCase.Item item;
		}

		public class Item
		{
		}

		private bool doThrow;

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			config.ObjectClass(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Holder)).
				CascadeOnUpdate(true);
			config.ObjectClass(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Holder)).
				CascadeOnDelete(true);
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			Store(new ExceptionInUpdatingCallbackCorruptionTestCase.Holder());
			Store(new ExceptionInUpdatingCallbackCorruptionTestCase.Item());
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Db4oSetupAfterStore()
		{
			EventRegistryFactory.ForObjectContainer(Db()).Updated += new System.EventHandler<Db4objects.Db4o.Events.ObjectInfoEventArgs>
				(new _IEventListener4_48(this).OnEvent);
		}

		private sealed class _IEventListener4_48
		{
			public _IEventListener4_48(ExceptionInUpdatingCallbackCorruptionTestCase _enclosing
				)
			{
				this._enclosing = _enclosing;
			}

			public void OnEvent(object sender, Db4objects.Db4o.Events.ObjectInfoEventArgs args
				)
			{
				if (this._enclosing.doThrow)
				{
					if (((ObjectInfoEventArgs)args).Info.GetObject().GetType().Equals(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Item
						)))
					{
						throw new Exception();
					}
				}
			}

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase _enclosing;
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void TestExceptionDuringItemUpdate()
		{
			ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder = ((ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)));
			ExceptionInUpdatingCallbackCorruptionTestCase.Item item = ((ExceptionInUpdatingCallbackCorruptionTestCase.Item
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Item)
				));
			holder.item = item;
			WithException(new _IBlock4_63(this, holder));
			CheckConsistencyFull();
		}

		private sealed class _IBlock4_63 : IBlock4
		{
			public _IBlock4_63(ExceptionInUpdatingCallbackCorruptionTestCase _enclosing, ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				 holder)
			{
				this._enclosing = _enclosing;
				this.holder = holder;
			}

			public void Run()
			{
				this._enclosing.Db().Store(holder, int.MaxValue);
			}

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase _enclosing;

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder;
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void TestExceptionDuringExistingListUpdate()
		{
			ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder = ((ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)));
			ExceptionInUpdatingCallbackCorruptionTestCase.Item item = ((ExceptionInUpdatingCallbackCorruptionTestCase.Item
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Item)
				));
			holder.list.Add(item);
			WithException(new _IBlock4_75(this, holder));
			CheckConsistencyFull();
		}

		private sealed class _IBlock4_75 : IBlock4
		{
			public _IBlock4_75(ExceptionInUpdatingCallbackCorruptionTestCase _enclosing, ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				 holder)
			{
				this._enclosing = _enclosing;
				this.holder = holder;
			}

			public void Run()
			{
				this._enclosing.Db().Store(holder, int.MaxValue);
			}

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase _enclosing;

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder;
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void TestExceptionDuringNewListUpdate()
		{
			ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder = ((ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				)));
			ExceptionInUpdatingCallbackCorruptionTestCase.Item item = ((ExceptionInUpdatingCallbackCorruptionTestCase.Item
				)RetrieveOnlyInstance(typeof(ExceptionInUpdatingCallbackCorruptionTestCase.Item)
				));
			holder.list = new ArrayList();
			holder.list.Add(item);
			WithException(new _IBlock4_88(this, holder));
			CheckConsistencyFull();
		}

		private sealed class _IBlock4_88 : IBlock4
		{
			public _IBlock4_88(ExceptionInUpdatingCallbackCorruptionTestCase _enclosing, ExceptionInUpdatingCallbackCorruptionTestCase.Holder
				 holder)
			{
				this._enclosing = _enclosing;
				this.holder = holder;
			}

			public void Run()
			{
				this._enclosing.Db().Store(holder, int.MaxValue);
			}

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase _enclosing;

			private readonly ExceptionInUpdatingCallbackCorruptionTestCase.Holder holder;
		}

		private void WithException(IBlock4 block)
		{
			doThrow = true;
			try
			{
				block.Run();
			}
			catch (Exception)
			{
			}
			finally
			{
				doThrow = false;
			}
		}

		/// <exception cref="System.Exception"></exception>
		private void CheckConsistencyFull()
		{
			CheckConsistency();
			Commit();
			CheckConsistency();
			Reopen();
			CheckConsistency();
		}

		private void CheckConsistency()
		{
			ConsistencyReport report = new ConsistencyChecker((LocalObjectContainer)Container
				()).CheckSlotConsistency();
			if (!report.Consistent())
			{
				Assert.Fail(report.ToString());
			}
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/events/ExceptionInUpdatingCallbackCorruptionTestCase.java
/* Copyright (C) 2004 - 2010  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.events;

import java.util.*;

import com.db4o.config.*;
import com.db4o.consistency.*;
import com.db4o.events.*;
import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.slots.*;

import db4ounit.*;
import db4ounit.extensions.*;
import db4ounit.extensions.fixtures.*;

/**
 * @exclude
 */
public class ExceptionInUpdatingCallbackCorruptionTestCase extends AbstractDb4oTestCase implements OptOutMultiSession {
	
	public static void main(String[] args) {
		new ExceptionInUpdatingCallbackCorruptionTestCase().runSolo();
	}
	
	public static class Holder {
		public List<Item> list = new ArrayList<Item>();
		public Item item;
	}
	
	public static class Item {
	}
	
	private boolean doThrow;
	
	@Override
	protected void configure(Configuration config) throws Exception {
		config.objectClass(Holder.class).cascadeOnUpdate(true);
		config.objectClass(Holder.class).cascadeOnDelete(true);
	}
	
	@Override
	protected void store() throws Exception {
		store(new Holder());
		store(new Item());
	}

	@Override
	protected void db4oSetupAfterStore() throws Exception {
		EventRegistryFactory.forObjectContainer(db()).updated().addListener(new EventListener4<ObjectInfoEventArgs>() {
			@Override
			public void onEvent(Event4<ObjectInfoEventArgs> e, ObjectInfoEventArgs args) {				
				if(doThrow){
					if(args.info().getObject().getClass().equals(Item.class)){
						throw new RuntimeException();
					}
				}
			}
		});
	}

	public void testExceptionDuringItemUpdate() throws Exception{
		final Holder holder = retrieveOnlyInstance(Holder.class);
		final Item item = retrieveOnlyInstance(Item.class);
		holder.item = item;
		withException(new Block4() {
			@Override
			public void run() {
				db().store(holder, Integer.MAX_VALUE);
			}
		});
		checkConsistencyFull();		
	}

	public void testExceptionDuringExistingListUpdate() throws Exception{
		final Holder holder = retrieveOnlyInstance(Holder.class);
		Item item = retrieveOnlyInstance(Item.class);
		holder.list.add(item);
		withException(new Block4() {
			@Override
			public void run() {
				db().store(holder, Integer.MAX_VALUE);
			}
		});
		checkConsistencyFull();		
	}

	public void testExceptionDuringNewListUpdate() throws Exception{
		final Holder holder = retrieveOnlyInstance(Holder.class);
		Item item = retrieveOnlyInstance(Item.class);
		holder.list = new ArrayList<Item>();
		holder.list.add(item);
		withException(new Block4() {
			@Override
			public void run() {
				db().store(holder, Integer.MAX_VALUE);
			}
		});
		checkConsistencyFull();
	}

	private void withException(Block4 block) {
		doThrow = true;
		try {
			block.run();
		}
		catch(RuntimeException exc) {
		}
		finally {
			doThrow = false;
		}
	}
	
	private void checkConsistencyFull() throws Exception {
		checkConsistency();
		commit();
		checkConsistency();
		reopen();
		checkConsistency();
	}

	private void checkConsistency() {
		ConsistencyReport report = new ConsistencyChecker((LocalObjectContainer) container()).checkSlotConsistency();
		if(!report.consistent()) {
			Assert.fail(report.toString());
		}
	}

}

